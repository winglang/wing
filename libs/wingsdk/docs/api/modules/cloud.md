[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / cloud

# Namespace: cloud

## Table of contents

### Enumerations

- [BucketInflightMethods](../enums/cloud.BucketInflightMethods.md)
- [FunctionInflightMethods](../enums/cloud.FunctionInflightMethods.md)
- [QueueInflightMethods](../enums/cloud.QueueInflightMethods.md)

### Classes

- [Bucket](../classes/cloud.Bucket.md)
- [BucketBase](../classes/cloud.BucketBase.md)
- [Function](../classes/cloud.Function.md)
- [FunctionBase](../classes/cloud.FunctionBase.md)
- [Queue](../classes/cloud.Queue.md)
- [QueueBase](../classes/cloud.QueueBase.md)
- [Resource](../classes/cloud.Resource.md)

### Interfaces

- [BucketProps](../interfaces/cloud.BucketProps.md)
- [FunctionProps](../interfaces/cloud.FunctionProps.md)
- [IBucketClient](../interfaces/cloud.IBucketClient.md)
- [IFunctionClient](../interfaces/cloud.IFunctionClient.md)
- [IQueueClient](../interfaces/cloud.IQueueClient.md)
- [QueueOnMessageProps](../interfaces/cloud.QueueOnMessageProps.md)
- [QueueProps](../interfaces/cloud.QueueProps.md)

### Variables

- [BUCKET\_TYPE](cloud.md#bucket_type)
- [FUNCTION\_TYPE](cloud.md#function_type)
- [QUEUE\_TYPE](cloud.md#queue_type)

## Variables

### BUCKET\_TYPE

• `Const` **BUCKET\_TYPE**: ``"wingsdk.cloud.Bucket"``

Global identifier for `Bucket`.

#### Defined in

[src/cloud/bucket.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/bucket.ts#L9)

___

### FUNCTION\_TYPE

• `Const` **FUNCTION\_TYPE**: ``"wingsdk.cloud.Function"``

Global identifier for `Function`.

#### Defined in

[src/cloud/function.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/function.ts#L9)

___

### QUEUE\_TYPE

• `Const` **QUEUE\_TYPE**: ``"wingsdk.cloud.Queue"``

Global identifier for `Queue`.

#### Defined in

[src/cloud/queue.ts:10](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/queue.ts#L10)
