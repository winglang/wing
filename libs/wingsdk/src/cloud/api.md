---
title: Api
id: api
description: A built-in resource for creating HTTP endpoints in the cloud.
keywords:
  [
    Wing reference,
    Wing language,
    language,
    Wing standard library,
    Wing programming language,
    APIs,
    Endpoints,
    HTTP endpoint,
  ]
sidebar_position: 1
---

The `cloud.Api` resource represents a collection of HTTP endpoints that can be invoked by clients over the internet.
APIs often serve as the front door for applications to access data, business logic, or functionality from your backend services.

The `Api` resource models an endpoint as a collection of routes, each mapped to an event handler function.
A route is a combination of a path, like `"/users/:userid"` and a set of HTTP methods, like `GET`, `POST`, or `DELETE`.
When a client invokes a route, the corresponding event handler function executes.

## Usage

### REST API

The following example shows a complete REST API implementation using `cloud.Api`, `ex.Table` & `cloud.Counter`

```ts playground example
bring cloud;
bring ex;

let api = new cloud.Api();
// Used for generating unique id
let counter = new cloud.Counter();
// our employee database
let db = new ex.Table(
  name: "employees",
  primaryKey: "id",
  columns: {
    "name" => ex.ColumnType.STRING
  }
);

api.get("/employees", inflight (request: cloud.ApiRequest): cloud.ApiResponse => {
  let result = MutJson [];
  let var i = 0;
  for employee in db.list() {
    result.setAt(i, employee);
    i = i + 1;
  }
  return cloud.ApiResponse {
    status: 200,
    body: Json.stringify(result)
  };
});


api.get("/employees/:id", inflight (request: cloud.ApiRequest): cloud.ApiResponse => {
  let employee = db.get(request.vars.get("id"));
  return cloud.ApiResponse {
    status: 200,
    body: Json.stringify(employee)
  };
});

api.post("/employees", inflight (request: cloud.ApiRequest): cloud.ApiResponse => {
   if let body = request.body {
    let employeeData = Json.parse(body);
    let id = "{counter.inc()}";
    db.insert(id, employeeData);
    return cloud.ApiResponse {
      status: 201,
      body: id
    };
   }
});

api.put("/employees/:id", inflight (request: cloud.ApiRequest): cloud.ApiResponse => {
  if let body = request.body {
    let employeeData = Json.parse(body);
    let id = request.vars.get("id");
    db.update(id, employeeData);
    return cloud.ApiResponse {
      status: 200,
      body: Json.stringify(employeeData)
    };
  }
});

api.delete("/employees/:id", inflight (request: cloud.ApiRequest): cloud.ApiResponse => {
  let id = request.vars.get("id");
  db.delete(id);
  return cloud.ApiResponse {
    status: 204
  };
});
```

## Target-specific details

### Simulator (`sim`)

The sim implementation of `cloud.Api` uses [nodejs express](https://expressjs.com/).

### AWS (`tf-aws` and `awscdk`)

The AWS implementation of `cloud.Api` uses [Amazon API Gateway](https://aws.amazon.com/api-gateway/).

### Azure (`tf-azure`)

ðŸš§ Not supported yet (tracking issue: [#625](https://github.com/winglang/wing/issues/625))

### GCP (`tf-gcp`)

ðŸš§ Not supported yet (tracking issue: [#624](https://github.com/winglang/wing/issues/624))
