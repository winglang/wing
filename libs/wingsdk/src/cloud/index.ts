export * from "./api";
export * from "./bucket";
export * from "./counter";
export * from "./domain";
export * from "./endpoint";
export * from "./function";
export * from "./on-deploy";
export * from "./queue";
export * from "./schedule";
export * from "./secret";
export * from "./service";
export * from "./topic";
export * from "./website";
