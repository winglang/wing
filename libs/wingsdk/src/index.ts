// this file must only export the standard library types and not target-specific types.
export * as cloud from "./cloud";
export * as core from "./core";
export * as ex from "./ex";
export * as expect from "./expect";
export * as fs from "./fs";
export * as helpers from "./helpers";
export * as http from "./http";
export * as math from "./math";
export * as platform from "./platform";
export * as regex from "./regex";
export * as aws from "./shared-aws";
export * as simulator from "./simulator";
export * as std from "./std";
export * as sim from "./target-sim";
export * as ui from "./ui";
export * as util from "./util";
