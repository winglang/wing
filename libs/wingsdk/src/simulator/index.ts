export * from "./client";
export * from "./simulator";
export * from "./tree";
