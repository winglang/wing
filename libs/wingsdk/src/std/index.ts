export * from "./array";
export * from "./bool";
export * from "./duration";
export * from "./json";
export * from "./map";
export * from "./number";
export * from "./range";
export * from "./resource";
export * from "./set";
export * from "./string";
export * from "./test";
export * from "./util";
