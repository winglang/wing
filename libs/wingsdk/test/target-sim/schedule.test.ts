import { test, expect } from "vitest";
import * as cloud from "../../src/cloud";
import { Testing } from "../../src/simulator";
import { Duration } from "../../src/std";
import { SimApp } from "../sim-app";

const INFLIGHT_CODE = `
async handle(message) {
console.log("Hello from schedule!");
}`;

test("create a schedule", async () => {
  // GIVEN
  const app = new SimApp();
  const cron = "*/1 * * * *";
  new cloud.Schedule(app, "my_schedule", { cron });
  const s = await app.startSimulator();

  // THEN
  expect(s.getResourceConfig("/my_schedule")).toEqual({
    attrs: {
      handle: expect.any(String),
    },
    path: "root/my_schedule",
    addr: expect.any(String),
    policy: [],
    props: {
      cronExpression: cron,
    },
    type: cloud.SCHEDULE_FQN,
  });
  await s.stop();

  // THEN
  expect(app.snapshot()).toMatchSnapshot();
});

test("schedule with one task with cron", async () => {
  // GIVEN
  const app = new SimApp();
  const handler = Testing.makeHandler(INFLIGHT_CODE);
  const schedule = new cloud.Schedule(app, "my_schedule", {
    cron: "* * * * *",
  });
  schedule.onTick(handler);

  const s = await app.startSimulator();
  // WHEN

  await s.stop();

  // THEN
  expect(app.snapshot()).toMatchSnapshot();
});

test("schedule with one task using rate of 10m", async () => {
  // GIVEN
  const app = new SimApp();
  const handler = Testing.makeHandler(INFLIGHT_CODE);
  const schedule = new cloud.Schedule(app, "my_schedule", {
    rate: Duration.fromMinutes(10),
  });
  const expectedCron = "*/10 * * * *"; // every 10 minutes cron expression
  schedule.onTick(handler);
  const s = await app.startSimulator();

  // THEN
  expect(app.snapshot()).toMatchSnapshot();
  expect(s.getResourceConfig("/my_schedule")).toEqual({
    attrs: {
      handle: expect.any(String),
    },
    path: "root/my_schedule",
    addr: expect.any(String),
    policy: [],
    props: {
      cronExpression: expectedCron,
    },
    type: cloud.SCHEDULE_FQN,
  });

  await s.stop();
});

test("schedule with one task using rate of 3h", async () => {
  // GIVEN
  const app = new SimApp();
  const handler = Testing.makeHandler(INFLIGHT_CODE);
  const schedule = new cloud.Schedule(app, "my_schedule", {
    rate: Duration.fromHours(3),
  });
  const expectedCron = "* */3 * * *"; // every 3 hours cron expression
  schedule.onTick(handler);
  const s = await app.startSimulator();

  // THEN
  expect(app.snapshot()).toMatchSnapshot();
  expect(s.getResourceConfig("/my_schedule")).toEqual({
    attrs: {
      handle: expect.any(String),
    },
    path: "root/my_schedule",
    addr: expect.any(String),
    policy: [],
    props: {
      cronExpression: expectedCron,
    },
    type: cloud.SCHEDULE_FQN,
  });

  await s.stop();
});
