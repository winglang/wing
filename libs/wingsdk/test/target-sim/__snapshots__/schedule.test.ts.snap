// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`create a schedule 1`] = `
{
  "connections.json": {
    "connections": [],
    "version": "connections-0.1",
  },
  "simulator.json": {
    "resources": [
      {
        "addr": "c8eac5abdbcb1c98c5525dcb3d0e38c0ccdfef4aed",
        "attrs": {},
        "path": "root/my_schedule",
        "props": {
          "cronExpression": "*/1 * * * ?",
        },
        "type": "@winglang/sdk.cloud.Schedule",
      },
    ],
    "sdkVersion": "0.0.0",
    "types": {
      "@winglang/sdk.cloud.Api": {
        "className": "Api",
        "sourcePath": "<ABSOLUTE PATH>/api.inflight.js",
      },
      "@winglang/sdk.cloud.Bucket": {
        "className": "Bucket",
        "sourcePath": "<ABSOLUTE PATH>/bucket.inflight.js",
      },
      "@winglang/sdk.cloud.Counter": {
        "className": "Counter",
        "sourcePath": "<ABSOLUTE PATH>/counter.inflight.js",
      },
      "@winglang/sdk.cloud.Domain": {
        "className": "Domain",
        "sourcePath": "<ABSOLUTE PATH>/domain.inflight.js",
      },
      "@winglang/sdk.cloud.Endpoint": {
        "className": "Endpoint",
        "sourcePath": "<ABSOLUTE PATH>/endpoint.inflight.js",
      },
      "@winglang/sdk.cloud.Function": {
        "className": "Function",
        "sourcePath": "<ABSOLUTE PATH>/function.inflight.js",
      },
      "@winglang/sdk.cloud.OnDeploy": {
        "className": "OnDeploy",
        "sourcePath": "<ABSOLUTE PATH>/on-deploy.inflight.js",
      },
      "@winglang/sdk.cloud.Queue": {
        "className": "Queue",
        "sourcePath": "<ABSOLUTE PATH>/queue.inflight.js",
      },
      "@winglang/sdk.cloud.Schedule": {
        "className": "Schedule",
        "sourcePath": "<ABSOLUTE PATH>/schedule.inflight.js",
      },
      "@winglang/sdk.cloud.Secret": {
        "className": "Secret",
        "sourcePath": "<ABSOLUTE PATH>/secret.inflight.js",
      },
      "@winglang/sdk.cloud.Service": {
        "className": "Service",
        "sourcePath": "<ABSOLUTE PATH>/service.inflight.js",
      },
      "@winglang/sdk.cloud.Topic": {
        "className": "Topic",
        "sourcePath": "<ABSOLUTE PATH>/topic.inflight.js",
      },
      "@winglang/sdk.cloud.Website": {
        "className": "Website",
        "sourcePath": "<ABSOLUTE PATH>/website.inflight.js",
      },
      "@winglang/sdk.ex.DynamodbTable": {
        "className": "DynamodbTable",
        "sourcePath": "<ABSOLUTE PATH>/dynamodb-table.inflight.js",
      },
      "@winglang/sdk.ex.ReactApp": {
        "className": "ReactApp",
        "sourcePath": "<ABSOLUTE PATH>/react-app.inflight.js",
      },
      "@winglang/sdk.ex.Redis": {
        "className": "Redis",
        "sourcePath": "<ABSOLUTE PATH>/redis.inflight.js",
      },
      "@winglang/sdk.ex.Table": {
        "className": "Table",
        "sourcePath": "<ABSOLUTE PATH>/table.inflight.js",
      },
      "@winglang/sdk.sim.EventMapping": {
        "className": "EventMapping",
        "sourcePath": "<ABSOLUTE PATH>/event-mapping.inflight.js",
      },
      "@winglang/sdk.sim.State": {
        "className": "State",
        "sourcePath": "<ABSOLUTE PATH>/state.inflight.js",
      },
      "@winglang/sdk.std.TestRunner": {
        "className": "TestRunner",
        "sourcePath": "<ABSOLUTE PATH>/test-runner.inflight.js",
      },
    },
  },
  "tree.json": {
    "tree": {
      "children": {
        "my_schedule": {
          "constructInfo": {
            "fqn": "constructs.Construct",
            "version": "10.2.70",
          },
          "display": {
            "description": "A cloud schedule to trigger events at regular intervals",
            "title": "Schedule",
          },
          "id": "my_schedule",
          "path": "root/my_schedule",
        },
      },
      "constructInfo": {
        "fqn": "constructs.Construct",
        "version": "10.2.70",
      },
      "id": "root",
      "path": "root",
    },
    "version": "tree-0.1",
  },
}
`;

exports[`schedule with one task using rate of 3h 1`] = `
{
  ".wing/ontick0_c83eddbe.js": "\\"use strict\\";
var $handler = undefined;
exports.handler = async function(event) {
  $handler = $handler ?? (new (require(\\"[REDACTED]/wingsdk/src/target-sim/schedule.ontick.inflight.js\\")).ScheduleOnTickHandlerClient({ handler: new ((function(){
return class Handler {
  constructor(clients) {
    for (const [name, client] of Object.entries(clients)) {
      this[name] = client;
    }
  }
  
async handle(message) {
console.log(\\"Hello from schedule!\\");
}
};
})())({

}), args: {} }));
  return await $handler.handle(event);
};",
  "connections.json": {
    "connections": [
      {
        "name": "onTick()",
        "source": "root/my_schedule",
        "target": "root/my_schedule/OnTick0",
      },
    ],
    "version": "connections-0.1",
  },
  "simulator.json": {
    "resources": [
      {
        "addr": "c83eddbe8512d15b84969a611d4b5ca9ce55e1f808",
        "attrs": {},
        "path": "root/my_schedule/OnTick0",
        "props": {
          "environmentVariables": {},
          "sourceCodeFile": ".wing/ontick0_c83eddbe.js",
          "sourceCodeLanguage": "javascript",
          "timeout": 60000,
        },
        "type": "@winglang/sdk.cloud.Function",
      },
      {
        "addr": "c8eac5abdbcb1c98c5525dcb3d0e38c0ccdfef4aed",
        "attrs": {},
        "path": "root/my_schedule",
        "props": {
          "cronExpression": "* */3 * * ?",
        },
        "type": "@winglang/sdk.cloud.Schedule",
      },
      {
        "addr": "c863abe75f0d01961369ad52fd7212846989919eaf",
        "attrs": {},
        "path": "root/my_schedule/OnTickMapping0",
        "props": {
          "publisher": "\${wsim#root/my_schedule#attrs.handle}",
          "subscriber": "\${wsim#root/my_schedule/OnTick0#attrs.handle}",
          "subscriptionProps": {},
        },
        "type": "@winglang/sdk.sim.EventMapping",
      },
    ],
    "sdkVersion": "0.0.0",
    "types": {
      "@winglang/sdk.cloud.Api": {
        "className": "Api",
        "sourcePath": "<ABSOLUTE PATH>/api.inflight.js",
      },
      "@winglang/sdk.cloud.Bucket": {
        "className": "Bucket",
        "sourcePath": "<ABSOLUTE PATH>/bucket.inflight.js",
      },
      "@winglang/sdk.cloud.Counter": {
        "className": "Counter",
        "sourcePath": "<ABSOLUTE PATH>/counter.inflight.js",
      },
      "@winglang/sdk.cloud.Domain": {
        "className": "Domain",
        "sourcePath": "<ABSOLUTE PATH>/domain.inflight.js",
      },
      "@winglang/sdk.cloud.Endpoint": {
        "className": "Endpoint",
        "sourcePath": "<ABSOLUTE PATH>/endpoint.inflight.js",
      },
      "@winglang/sdk.cloud.Function": {
        "className": "Function",
        "sourcePath": "<ABSOLUTE PATH>/function.inflight.js",
      },
      "@winglang/sdk.cloud.OnDeploy": {
        "className": "OnDeploy",
        "sourcePath": "<ABSOLUTE PATH>/on-deploy.inflight.js",
      },
      "@winglang/sdk.cloud.Queue": {
        "className": "Queue",
        "sourcePath": "<ABSOLUTE PATH>/queue.inflight.js",
      },
      "@winglang/sdk.cloud.Schedule": {
        "className": "Schedule",
        "sourcePath": "<ABSOLUTE PATH>/schedule.inflight.js",
      },
      "@winglang/sdk.cloud.Secret": {
        "className": "Secret",
        "sourcePath": "<ABSOLUTE PATH>/secret.inflight.js",
      },
      "@winglang/sdk.cloud.Service": {
        "className": "Service",
        "sourcePath": "<ABSOLUTE PATH>/service.inflight.js",
      },
      "@winglang/sdk.cloud.Topic": {
        "className": "Topic",
        "sourcePath": "<ABSOLUTE PATH>/topic.inflight.js",
      },
      "@winglang/sdk.cloud.Website": {
        "className": "Website",
        "sourcePath": "<ABSOLUTE PATH>/website.inflight.js",
      },
      "@winglang/sdk.ex.DynamodbTable": {
        "className": "DynamodbTable",
        "sourcePath": "<ABSOLUTE PATH>/dynamodb-table.inflight.js",
      },
      "@winglang/sdk.ex.ReactApp": {
        "className": "ReactApp",
        "sourcePath": "<ABSOLUTE PATH>/react-app.inflight.js",
      },
      "@winglang/sdk.ex.Redis": {
        "className": "Redis",
        "sourcePath": "<ABSOLUTE PATH>/redis.inflight.js",
      },
      "@winglang/sdk.ex.Table": {
        "className": "Table",
        "sourcePath": "<ABSOLUTE PATH>/table.inflight.js",
      },
      "@winglang/sdk.sim.EventMapping": {
        "className": "EventMapping",
        "sourcePath": "<ABSOLUTE PATH>/event-mapping.inflight.js",
      },
      "@winglang/sdk.sim.State": {
        "className": "State",
        "sourcePath": "<ABSOLUTE PATH>/state.inflight.js",
      },
      "@winglang/sdk.std.TestRunner": {
        "className": "TestRunner",
        "sourcePath": "<ABSOLUTE PATH>/test-runner.inflight.js",
      },
    },
  },
  "tree.json": {
    "tree": {
      "children": {
        "my_schedule": {
          "children": {
            "OnTick0": {
              "constructInfo": {
                "fqn": "constructs.Construct",
                "version": "10.2.70",
              },
              "display": {
                "description": "A cloud function (FaaS)",
                "sourceModule": "@winglang/sdk",
                "title": "onTick()",
              },
              "id": "OnTick0",
              "path": "root/my_schedule/OnTick0",
            },
            "OnTickMapping0": {
              "constructInfo": {
                "fqn": "constructs.Construct",
                "version": "10.2.70",
              },
              "display": {
                "hidden": true,
              },
              "id": "OnTickMapping0",
              "path": "root/my_schedule/OnTickMapping0",
            },
          },
          "constructInfo": {
            "fqn": "constructs.Construct",
            "version": "10.2.70",
          },
          "display": {
            "description": "A cloud schedule to trigger events at regular intervals",
            "title": "Schedule",
          },
          "id": "my_schedule",
          "path": "root/my_schedule",
        },
      },
      "constructInfo": {
        "fqn": "constructs.Construct",
        "version": "10.2.70",
      },
      "id": "root",
      "path": "root",
    },
    "version": "tree-0.1",
  },
}
`;

exports[`schedule with one task using rate of 10m 1`] = `
{
  ".wing/ontick0_c83eddbe.js": "\\"use strict\\";
var $handler = undefined;
exports.handler = async function(event) {
  $handler = $handler ?? (new (require(\\"[REDACTED]/wingsdk/src/target-sim/schedule.ontick.inflight.js\\")).ScheduleOnTickHandlerClient({ handler: new ((function(){
return class Handler {
  constructor(clients) {
    for (const [name, client] of Object.entries(clients)) {
      this[name] = client;
    }
  }
  
async handle(message) {
console.log(\\"Hello from schedule!\\");
}
};
})())({

}), args: {} }));
  return await $handler.handle(event);
};",
  "connections.json": {
    "connections": [
      {
        "name": "onTick()",
        "source": "root/my_schedule",
        "target": "root/my_schedule/OnTick0",
      },
    ],
    "version": "connections-0.1",
  },
  "simulator.json": {
    "resources": [
      {
        "addr": "c83eddbe8512d15b84969a611d4b5ca9ce55e1f808",
        "attrs": {},
        "path": "root/my_schedule/OnTick0",
        "props": {
          "environmentVariables": {},
          "sourceCodeFile": ".wing/ontick0_c83eddbe.js",
          "sourceCodeLanguage": "javascript",
          "timeout": 60000,
        },
        "type": "@winglang/sdk.cloud.Function",
      },
      {
        "addr": "c8eac5abdbcb1c98c5525dcb3d0e38c0ccdfef4aed",
        "attrs": {},
        "path": "root/my_schedule",
        "props": {
          "cronExpression": "*/10 * * * ?",
        },
        "type": "@winglang/sdk.cloud.Schedule",
      },
      {
        "addr": "c863abe75f0d01961369ad52fd7212846989919eaf",
        "attrs": {},
        "path": "root/my_schedule/OnTickMapping0",
        "props": {
          "publisher": "\${wsim#root/my_schedule#attrs.handle}",
          "subscriber": "\${wsim#root/my_schedule/OnTick0#attrs.handle}",
          "subscriptionProps": {},
        },
        "type": "@winglang/sdk.sim.EventMapping",
      },
    ],
    "sdkVersion": "0.0.0",
    "types": {
      "@winglang/sdk.cloud.Api": {
        "className": "Api",
        "sourcePath": "<ABSOLUTE PATH>/api.inflight.js",
      },
      "@winglang/sdk.cloud.Bucket": {
        "className": "Bucket",
        "sourcePath": "<ABSOLUTE PATH>/bucket.inflight.js",
      },
      "@winglang/sdk.cloud.Counter": {
        "className": "Counter",
        "sourcePath": "<ABSOLUTE PATH>/counter.inflight.js",
      },
      "@winglang/sdk.cloud.Domain": {
        "className": "Domain",
        "sourcePath": "<ABSOLUTE PATH>/domain.inflight.js",
      },
      "@winglang/sdk.cloud.Endpoint": {
        "className": "Endpoint",
        "sourcePath": "<ABSOLUTE PATH>/endpoint.inflight.js",
      },
      "@winglang/sdk.cloud.Function": {
        "className": "Function",
        "sourcePath": "<ABSOLUTE PATH>/function.inflight.js",
      },
      "@winglang/sdk.cloud.OnDeploy": {
        "className": "OnDeploy",
        "sourcePath": "<ABSOLUTE PATH>/on-deploy.inflight.js",
      },
      "@winglang/sdk.cloud.Queue": {
        "className": "Queue",
        "sourcePath": "<ABSOLUTE PATH>/queue.inflight.js",
      },
      "@winglang/sdk.cloud.Schedule": {
        "className": "Schedule",
        "sourcePath": "<ABSOLUTE PATH>/schedule.inflight.js",
      },
      "@winglang/sdk.cloud.Secret": {
        "className": "Secret",
        "sourcePath": "<ABSOLUTE PATH>/secret.inflight.js",
      },
      "@winglang/sdk.cloud.Service": {
        "className": "Service",
        "sourcePath": "<ABSOLUTE PATH>/service.inflight.js",
      },
      "@winglang/sdk.cloud.Topic": {
        "className": "Topic",
        "sourcePath": "<ABSOLUTE PATH>/topic.inflight.js",
      },
      "@winglang/sdk.cloud.Website": {
        "className": "Website",
        "sourcePath": "<ABSOLUTE PATH>/website.inflight.js",
      },
      "@winglang/sdk.ex.DynamodbTable": {
        "className": "DynamodbTable",
        "sourcePath": "<ABSOLUTE PATH>/dynamodb-table.inflight.js",
      },
      "@winglang/sdk.ex.ReactApp": {
        "className": "ReactApp",
        "sourcePath": "<ABSOLUTE PATH>/react-app.inflight.js",
      },
      "@winglang/sdk.ex.Redis": {
        "className": "Redis",
        "sourcePath": "<ABSOLUTE PATH>/redis.inflight.js",
      },
      "@winglang/sdk.ex.Table": {
        "className": "Table",
        "sourcePath": "<ABSOLUTE PATH>/table.inflight.js",
      },
      "@winglang/sdk.sim.EventMapping": {
        "className": "EventMapping",
        "sourcePath": "<ABSOLUTE PATH>/event-mapping.inflight.js",
      },
      "@winglang/sdk.sim.State": {
        "className": "State",
        "sourcePath": "<ABSOLUTE PATH>/state.inflight.js",
      },
      "@winglang/sdk.std.TestRunner": {
        "className": "TestRunner",
        "sourcePath": "<ABSOLUTE PATH>/test-runner.inflight.js",
      },
    },
  },
  "tree.json": {
    "tree": {
      "children": {
        "my_schedule": {
          "children": {
            "OnTick0": {
              "constructInfo": {
                "fqn": "constructs.Construct",
                "version": "10.2.70",
              },
              "display": {
                "description": "A cloud function (FaaS)",
                "sourceModule": "@winglang/sdk",
                "title": "onTick()",
              },
              "id": "OnTick0",
              "path": "root/my_schedule/OnTick0",
            },
            "OnTickMapping0": {
              "constructInfo": {
                "fqn": "constructs.Construct",
                "version": "10.2.70",
              },
              "display": {
                "hidden": true,
              },
              "id": "OnTickMapping0",
              "path": "root/my_schedule/OnTickMapping0",
            },
          },
          "constructInfo": {
            "fqn": "constructs.Construct",
            "version": "10.2.70",
          },
          "display": {
            "description": "A cloud schedule to trigger events at regular intervals",
            "title": "Schedule",
          },
          "id": "my_schedule",
          "path": "root/my_schedule",
        },
      },
      "constructInfo": {
        "fqn": "constructs.Construct",
        "version": "10.2.70",
      },
      "id": "root",
      "path": "root",
    },
    "version": "tree-0.1",
  },
}
`;

exports[`schedule with one task with cron 1`] = `
{
  ".wing/ontick0_c83eddbe.js": "\\"use strict\\";
var $handler = undefined;
exports.handler = async function(event) {
  $handler = $handler ?? (new (require(\\"[REDACTED]/wingsdk/src/target-sim/schedule.ontick.inflight.js\\")).ScheduleOnTickHandlerClient({ handler: new ((function(){
return class Handler {
  constructor(clients) {
    for (const [name, client] of Object.entries(clients)) {
      this[name] = client;
    }
  }
  
async handle(message) {
console.log(\\"Hello from schedule!\\");
}
};
})())({

}), args: {} }));
  return await $handler.handle(event);
};",
  "connections.json": {
    "connections": [
      {
        "name": "onTick()",
        "source": "root/my_schedule",
        "target": "root/my_schedule/OnTick0",
      },
    ],
    "version": "connections-0.1",
  },
  "simulator.json": {
    "resources": [
      {
        "addr": "c83eddbe8512d15b84969a611d4b5ca9ce55e1f808",
        "attrs": {},
        "path": "root/my_schedule/OnTick0",
        "props": {
          "environmentVariables": {},
          "sourceCodeFile": ".wing/ontick0_c83eddbe.js",
          "sourceCodeLanguage": "javascript",
          "timeout": 60000,
        },
        "type": "@winglang/sdk.cloud.Function",
      },
      {
        "addr": "c8eac5abdbcb1c98c5525dcb3d0e38c0ccdfef4aed",
        "attrs": {},
        "path": "root/my_schedule",
        "props": {
          "cronExpression": "* * * * ?",
        },
        "type": "@winglang/sdk.cloud.Schedule",
      },
      {
        "addr": "c863abe75f0d01961369ad52fd7212846989919eaf",
        "attrs": {},
        "path": "root/my_schedule/OnTickMapping0",
        "props": {
          "publisher": "\${wsim#root/my_schedule#attrs.handle}",
          "subscriber": "\${wsim#root/my_schedule/OnTick0#attrs.handle}",
          "subscriptionProps": {},
        },
        "type": "@winglang/sdk.sim.EventMapping",
      },
    ],
    "sdkVersion": "0.0.0",
    "types": {
      "@winglang/sdk.cloud.Api": {
        "className": "Api",
        "sourcePath": "<ABSOLUTE PATH>/api.inflight.js",
      },
      "@winglang/sdk.cloud.Bucket": {
        "className": "Bucket",
        "sourcePath": "<ABSOLUTE PATH>/bucket.inflight.js",
      },
      "@winglang/sdk.cloud.Counter": {
        "className": "Counter",
        "sourcePath": "<ABSOLUTE PATH>/counter.inflight.js",
      },
      "@winglang/sdk.cloud.Domain": {
        "className": "Domain",
        "sourcePath": "<ABSOLUTE PATH>/domain.inflight.js",
      },
      "@winglang/sdk.cloud.Endpoint": {
        "className": "Endpoint",
        "sourcePath": "<ABSOLUTE PATH>/endpoint.inflight.js",
      },
      "@winglang/sdk.cloud.Function": {
        "className": "Function",
        "sourcePath": "<ABSOLUTE PATH>/function.inflight.js",
      },
      "@winglang/sdk.cloud.OnDeploy": {
        "className": "OnDeploy",
        "sourcePath": "<ABSOLUTE PATH>/on-deploy.inflight.js",
      },
      "@winglang/sdk.cloud.Queue": {
        "className": "Queue",
        "sourcePath": "<ABSOLUTE PATH>/queue.inflight.js",
      },
      "@winglang/sdk.cloud.Schedule": {
        "className": "Schedule",
        "sourcePath": "<ABSOLUTE PATH>/schedule.inflight.js",
      },
      "@winglang/sdk.cloud.Secret": {
        "className": "Secret",
        "sourcePath": "<ABSOLUTE PATH>/secret.inflight.js",
      },
      "@winglang/sdk.cloud.Service": {
        "className": "Service",
        "sourcePath": "<ABSOLUTE PATH>/service.inflight.js",
      },
      "@winglang/sdk.cloud.Topic": {
        "className": "Topic",
        "sourcePath": "<ABSOLUTE PATH>/topic.inflight.js",
      },
      "@winglang/sdk.cloud.Website": {
        "className": "Website",
        "sourcePath": "<ABSOLUTE PATH>/website.inflight.js",
      },
      "@winglang/sdk.ex.DynamodbTable": {
        "className": "DynamodbTable",
        "sourcePath": "<ABSOLUTE PATH>/dynamodb-table.inflight.js",
      },
      "@winglang/sdk.ex.ReactApp": {
        "className": "ReactApp",
        "sourcePath": "<ABSOLUTE PATH>/react-app.inflight.js",
      },
      "@winglang/sdk.ex.Redis": {
        "className": "Redis",
        "sourcePath": "<ABSOLUTE PATH>/redis.inflight.js",
      },
      "@winglang/sdk.ex.Table": {
        "className": "Table",
        "sourcePath": "<ABSOLUTE PATH>/table.inflight.js",
      },
      "@winglang/sdk.sim.EventMapping": {
        "className": "EventMapping",
        "sourcePath": "<ABSOLUTE PATH>/event-mapping.inflight.js",
      },
      "@winglang/sdk.sim.State": {
        "className": "State",
        "sourcePath": "<ABSOLUTE PATH>/state.inflight.js",
      },
      "@winglang/sdk.std.TestRunner": {
        "className": "TestRunner",
        "sourcePath": "<ABSOLUTE PATH>/test-runner.inflight.js",
      },
    },
  },
  "tree.json": {
    "tree": {
      "children": {
        "my_schedule": {
          "children": {
            "OnTick0": {
              "constructInfo": {
                "fqn": "constructs.Construct",
                "version": "10.2.70",
              },
              "display": {
                "description": "A cloud function (FaaS)",
                "sourceModule": "@winglang/sdk",
                "title": "onTick()",
              },
              "id": "OnTick0",
              "path": "root/my_schedule/OnTick0",
            },
            "OnTickMapping0": {
              "constructInfo": {
                "fqn": "constructs.Construct",
                "version": "10.2.70",
              },
              "display": {
                "hidden": true,
              },
              "id": "OnTickMapping0",
              "path": "root/my_schedule/OnTickMapping0",
            },
          },
          "constructInfo": {
            "fqn": "constructs.Construct",
            "version": "10.2.70",
          },
          "display": {
            "description": "A cloud schedule to trigger events at regular intervals",
            "title": "Schedule",
          },
          "id": "my_schedule",
          "path": "root/my_schedule",
        },
      },
      "constructInfo": {
        "fqn": "constructs.Construct",
        "version": "10.2.70",
      },
      "id": "root",
      "path": "root",
    },
    "version": "tree-0.1",
  },
}
`;
